GDJS/GDJS/Extensions/Builtin/CommonInstructionsExtension.cpp

```diff
@@ -101,6 +103,76 @@ CommonInstructionsExtension::CommonInstructionsExtension() {
         return outputCode;
       });
 
+  GetAllEvents()["BuiltinCommonInstructions::Async"].SetCodeGenerator(
+      [](gd::BaseEvent& event_,
+         gd::EventsCodeGenerator& codeGenerator,
+         gd::EventsCodeGenerationContext& parentContext) {
+        gd::AsyncEvent& event = dynamic_cast<gd::AsyncEvent&>(event_);
+
+        const gd::String callbackFunctionName =
+            "asyncCallback" +
+            gd::String::From(codeGenerator.GenerateSingleUsageUniqueIdFor(
+                &event.GetInstruction()));
+
+        // Generate callback code
+        gd::EventsCodeGenerationContext callbackContext;
+        callbackContext.InheritsAsAsyncCallbackFrom(parentContext);
+
+        // Generate actions
+        gd::String actionsCode = codeGenerator.GenerateActionsListCode(
+            event.GetActions(), callbackContext);
+
+        // Generate subevents
+        if (event.HasSubEvents())  // Sub events
+        {
+          actionsCode += "\n{ //Subevents\n";
+          actionsCode += codeGenerator.GenerateEventsListCode(
+              event.GetSubEvents(), callbackContext);
+          actionsCode += "} //End of subevents\n";
+        }
+
+        // Compose the callback function and add outside main
+        const gd::String actionsDeclarationsCode =
+            codeGenerator.GenerateObjectsDeclarationCode(callbackContext);
+        const gd::String callbackCode = "function " + callbackFunctionName +
+                                        "(runtimeScene, asyncObjectsList) {\n" +
+                                        actionsDeclarationsCode + actionsCode +
+                                        "}\n";
+        codeGenerator.AddCustomCodeOutsideMain(callbackCode);
+
+        // Generate code to backup the objects lists
+        gd::String objectsListsCode =
+            parentContext.IsAsync()
+                ? "asyncObjectsList = "
+                  "gdjs.LongLivedObjectsList.from(asyncObjectsList);\n"
+                : "const asyncObjectsList = new gdjs.LongLivedObjectsList();\n";
+        // The objects to be declared of the callback context tell all objects
+        // used in the callbacks so that they can be backed up now.
+        for (const gd::String& objectToBackup :
+             callbackContext.GetAllDeclaredObjectsAcrossChildren()) {
+          if (parentContext.ObjectAlreadyDeclared(objectToBackup))
+            objectsListsCode +=
+                "for (const obj of " +
+                codeGenerator.GetObjectListName(objectToBackup, parentContext) +
+                ") asyncObjectsList.addObject(obj);\n";
+        }
+
+        const gd::String callbackCallCode =
+            "(runtimeScene) => (" + callbackFunctionName + "(" +
+            dynamic_cast<gdjs::EventsCodeGenerator&>(codeGenerator)
```

Comment:
I've not verified if this call is valid, but we're missing in the case of an events function the eventsFunctionContext. See this generated code for example:

```
 function asyncCallback7006916(runtimeScene, asyncObjectsList) {
      {runtimeScene.getVariables().get("SuccessVariable").add((typeof eventsFunctionContext !== 'undefined' ? Number(eventsFunctionContext.getArgument("IncreaseValue")) || 0 : 0));
      }}
```

I'll push a test for this.

Reply:
Actually, this line is fine (it's properly passing the runtimeScene, the eventsFunctionContext and the asyncObjectsLists), this is the declaration of the asyncCallback that is not aware that it is an async callback (i.e: a "closure") that is inside an eventsFunction, and so will get this parameter too.